lndhub-go: integrate LndHub.go into Nix-Bitcoin

prusnak opened the pull request:
LndHub.go is an accounting wrapper for the Lightning Network. It provides separate accounts for end-users and LndHub compatible API while written in Go.

upstream: https://github.com/getAlby/lndhub.go

This is a draft PR to gather feedback, also any help about how to test this is welcome!

erikarvstedt commented:
Thanks, Pavol, lndhub seems low maintenance and a good fit for nix-bitcoin. If it supported clightning, I'd use it right away.

Why do you set `FEE_RESERVE` to `true` (default is `false`)?

prusnak commented:
> If it supported clightning, I'd use it right away.

CLN might be supported in the future, see https://github.com/getAlby/lndhub.go/pull/123

> Why do you set `FEE_RESERVE` to `true` (default is `false`)?

Fixed in 644b6324aff2250be84c1a5d28599ec85ae43a49

erikarvstedt commented:
[Initial fixups](https://github.com/erikarvstedt/nix-bitcoin/tree/lndhub-ea/) to make the service run.
The postgresql connection fails with:
```
Jul 21 13:12:16 nb-test lndhub.go[5251]: {"level":"fatal","time":"2022-07-21T13:12:16+02:00","message":"Error initializing db migrator: pgdriver: SASL: FATAL: password authentication failed for user \"lndhub-go\" (SQLSTATE=28P01)"}
```
`lndhub-go` should allow connections via the postgresql socket.

erikarvstedt commented:
I've removed all app-specific settings and moved them into settings.
If you think some of these settings deserve a dedicated NixOS option, we should add a [freeform module](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules).

prusnak commented:
Merged your fixups into my branch, added bfceb1e1db3d7d997b83d93881e3ba0bac674cdb which fixes the database name issue.

erikarvstedt commented:
[Fixups](https://github.com/erikarvstedt/nix-bitcoin/commits/7c92322b6f4aa212be12ff9507756bdc70d91866).
The postgres connection fails with:
```
Jul 21 14:40:23 nb-test lndhub.go[575]: {"level":"fatal","time":"2022-07-21T14:40:23+02:00","message":"Error initializing db migrator: pgdriver: SASL: FATAL: password authentication failed for user \"lndhub-go\" (SQLSTATE=28P01)"}
```
We can add a password via the secrets mechanism, but I'd really prefer a socket connection. This should be simple to add to lndhub.

prusnak commented:
> This should be simple to add to lndhub.

Maybe it is enough to just change the connection string to `postgres:///lndhubgo?host=/var/run/postgresql/` ?

erikarvstedt commented:
Ah, great. 
With `DATABASE_URI = "postgres:///lndhubgo?host=/run/postgresql";` connecting still fails:
```
Jul 21 16:03:10 nb-test lndhub.go[632]: {"level":"fatal","time":"2022-07-21T16:03:10+02:00","message":"Error initializing db migrator: dial unix /run/postgresql: connect: permission denied"}
```
But manually connecting via `sudo -u lndhub-go psql lndhubgo --host=/run/postgresql` works fine. Seems like a lndhub-go bug.

prusnak commented:
Maybe we need to specify also the user?

> `DATABASE_URI = "postgres:///lndhubgo?host=/run/postgresql&user=${cfg.user}";`

erikarvstedt commented:
No, this didn't help. Same with the equivalent `postgres://lndhub-go@/lndhubgo?host=/run/postgresql`

prusnak commented:
I figured it out. First, this PR has to be merged: https://github.com/getAlby/lndhub.go/pull/215 and then one can use the following connection string: `unix://${cfg.user}@lndhubgo/var/pgsql_socket/.s.PGSQL.5432?sslmode=disable`

prusnak commented:
When https://github.com/getAlby/lndhub.go/pull/218 is merged, we can remove the `xxd` shenanigans

jonasnick commented:
> Thanks, Pavol, lndhub seems low maintenance and a good fit for nix-bitcoin. If it supported clightning, I'd use it right away.

Agreed!

prusnak commented:
Rebased

prusnak commented:
lndhub-go 0.10.0 which contains the required fixes in https://github.com/getAlby/lndhub.go/pull/215 and https://github.com/getAlby/lndhub.go/pull/218 has been released and merged to nixpkgs via https://github.com/NixOS/nixpkgs/pull/185649

prusnak commented:
Rebased after https://github.com/fort-nix/nix-bitcoin/pull/537 has been merged. Removing draft status.

erikarvstedt commented:
[Fixups](https://github.com/erikarvstedt/nix-bitcoin/commits/lndhub-ea-3).

prusnak commented:
> [Fixups](https://github.com/erikarvstedt/nix-bitcoin/commits/lndhub-ea-3).

Merged into the PR

erikarvstedt commented:
Can you squash 'em?

erikarvstedt commented on modules/lndhub-go.nix:
```diff
@@ -0,0 +1,131 @@
+      BRANDING_TITLE = "LndHub.go - Nix-Bitcoin";
+      BRANDING_DESC = "Accounting wrapper for the Lightning Network";
+      BRANDING_URL = "https://nixbitcoin.org";
+      BRANDING_LOGO = "https://nixbitcoin.org/files/nix-bitcoin-logo-text.png";
```
This logo is too large because it's embedded without scaling. 
Maybe add scaling to the upstream layout or remove the custom logo.

prusnak replied:
Scaling added to upstream in https://github.com/getAlby/lndhub.go/pull/225

erikarvstedt replied:
This might take a while. Let's use [this branch](https://github.com/erikarvstedt/nix-bitcoin/commits/lndhub-ea-4) which contains the squashed fixups and the logo layout patch.

prusnak replied:
Force pushed `lndhub-ea-4` into this PR

erikarvstedt reviewed (approved):
ACK cff12ba68eea91970ffe75b5eccfe429501c1159

erikarvstedt reviewed (changes requested):
[Final fixup](https://github.com/erikarvstedt/nix-bitcoin/commits/lndhub-ea-5)

prusnak commented:
> Final fixup

applied to a03a0b79e51c7967044be09af5a2710a1e036a72

erikarvstedt reviewed (dismissed):
ACK 6a195f6102ede508e1555bb0cbe0d5b2aaf33764

prusnak commented:
Dropped 6a195f6102ede508e1555bb0cbe0d5b2aaf33764 since https://github.com/NixOS/nixpkgs/pull/204769 has been merged and no longer required.

Rebased ...
